File: test/phase_advance_tuning_demo.jl
# Example of using JuTrack to tune the ESR lattice for a specific phase advance between two crab cavities
# Warning: This code is for demonstration purposes. Simply tuning the quadrupoles may result in unstable solutions.
using Pkg
Pkg.activate("."); Pkg.instantiate() # change "." to your path of JuTrack.jl
using JuTrack
using Serialization
using Printf
set_tps_dim(7)  # 7 variables for the TPSA
E0 = 17.846262619763e9

# Load the ESR main lattice 
ESR_crab = deserialize("src/demo/ESR/esr_main.jls")
ESR_crab = Number2TPSAD(ESR_crab) # Convert to TPSA format

function Q_perturb(ESR_crab)
    for i in eachindex(ESR_crab)
        if ESR_crab[i] isa KQUAD
            k1 = ESR_crab[i].k1
            k1 = k1 * (1 + 0.001 * randn_approx())
            new_KQ = KQUAD(name=ESR_crab[i].name, len=ESR_crab[i].len, k1=k1)
            ESR_crab[i] = new_KQ
        end
    end
    return ESR_crab
end

ESR_perturb = Q_perturb(ESR_crab) # Perturbed ESR lattice
idx = findelem(ESR_crab, CRABCAVITY)

function get_phase14_zero(x1, x2, x3, x4, x5, x6, x7)
    X = [x1, x2, x3, x4, x5, x6, x7]
    changed_idx = [9,13,17,23,27,31,5537] # indices of the quadrupoles to be changed
    E0 = 17.846262619763e9

    refpts = [i for i in 1:length(ESR_crab)]
    for i in 1:length(changed_idx)
        ESR_perturb[changed_idx[i]].k1 = X[i]
    end
    # refpts = [35, 5533, length(ESR_crab)]
    twi = twissring(ESR_perturb, 0.0, 0, refpts, E0=E0, m0=m_e)

    phase41 = twi[35].mux + twi[end].mux - twi[5533].mux
    return phase41 - 2*pi
end

function multi_val_op(x0, niter, step, RING)
    target = 0.015
    x0_vals = zeros(7, niter)
    goal_vals = []
    grad_vals = zeros(7, niter)
    g0 = get_phase14_zero(x0[1], x0[2], x0[3], x0[4], x0[5], x0[6], x0[7])
    for i in 1:niter
        grads, new_phase = Gradient(get_phase14_zero, x0, true)
        grad9 = grads[1]
        grad13 = grads[2]
        grad17 = grads[3]
        grad23 = grads[4]
        grad27 = grads[5]
        grad31 = grads[6]
        grad5537 = grads[7]

        x0[1] -= step * grad9[1]
        x0[2] -= step * grad13[1]
        x0[3] -= step * grad17[1]
        x0[4] -= step * grad23[1]
        x0[5] -= step * grad27[1]
        x0[6] -= step * grad31[1]
        x0[7] -= step * grad5537[1]

        println(@sprintf("init: %.6f now: %.6f at step %d, 
        k1 : %.6f, k2 : %.6f, k3 : %.6f, k4 : %.6f, k5 : %.6f, k6 : %.6f, k7 : %.6f", g0.val, new_phase, i,
            x0[1], x0[2], x0[3], x0[4], x0[5], x0[6], x0[7]))
        x0_vals[:, i] = x0
        push!(goal_vals, new_phase)
        grad_vals[:, i] = [grad9[1], grad13[1], grad17[1], grad23[1], grad27[1], grad31[1], grad5537[1]]
        if new_phase < target 
            println("tuning finished at step ", i)
            break
        end
    end
    return x0_vals, goal_vals, grad_vals
end

xinit = [-1e-6, 1e-6, -1e-6, 1e-6, -1e-6, 1e-6, -1e-6] # Initial guess for the quadrupole strengths
x0_vals, goal_vals, grad_vals = multi_val_op(xinit, 100, 1e-4, ESR_crab)

# using LaTeXStrings
# plot_steps = length(goal_vals)
# p1 = plot(1:plot_steps, x0_vals[1, 1:plot_steps], title = L"Evolution\ of\ k", xlabel = L"Iterations", ylabel = L"Strength (m^{-1})", label=L"k_1", line=:dash, marker=:circle,framestyle=:box)
# for i in 2:7
#     label_str = "k_{$i}"  
#     full_label = latexstring(label_str)
#     plot!(1:plot_steps, x0_vals[i, 1:plot_steps], label=full_label, line=:dash, marker=:circle)
# end
# p2 = plot(1:plot_steps, goal_vals[1:plot_steps], title = L"Evolution\ of\ \Delta \phi", xlabel = L"Iterations", 
#     ylabel = L"phase\ advance(rad)", legend = false, line=:dash, marker=:circle,framestyle=:box)
# p3 = plot(1:plot_steps, grad_vals[1, 1:plot_steps], title = L"Evolution\ of\ gradient", xlabel = L"Iterations", 
#     ylabel = L"\partial \frac{\Delta \phi}{\partial k}", label=L"k_1", line=:dash, marker=:circle,framestyle=:box)
# for i in 2:7
#     label_str = "k_{$i}"  
#     full_label = latexstring(label_str)
#     plot!(1:plot_steps, grad_vals[i, 1:plot_steps], label=full_label, line=:dash, marker=:circle)
# end
# plot(p1, p2, p3, layout = (3, 1), size=(800, 650))
# # savefig("plot1.png")


##########################################################
# Legacy code of using Enzyme for gradient calculation
##########################################################
# E0 = 17.846262619763e9

# ESR_crab = deserialize("src/demo/ESR/esr_main_linearquad.jls")

# function Q_perturb(ESR_crab)
#     for i in eachindex(ESR_crab)
#         if ESR_crab[i] isa KQUAD
#             k1 = ESR_crab[i].k1
#             k1 = k1 * (1 + 0.001 * randn_approx())
#             new_KQ = KQUAD(name=ESR_crab[i].name, len=ESR_crab[i].len, k1=k1)
#             ESR_crab[i] = new_KQ
#         end
#     end
#     return ESR_crab
# end

# ESR_perturb = Q_perturb(ESR_crab) # Perturbed ESR lattice

# idx = findelem(ESR_crab,CRABCAVITY)

# xinit = [-1e-6, 1e-6, -1e-6, 1e-6, -1e-6, 1e-6, -1e-6]

# zero_idx = [9,13,17,23,27,31,5537]
# function get_phase14_zero(x, ESR_perturb)
#     changed_idx = [9,13,17,23,27,31,5537]
#     L = 0.25
#     E0 = 17.846262619763e9
#     new_Q1 = QUAD(len=L, k1=x[1])
#     new_Q2 = QUAD(len=L, k1=x[2])
#     new_Q3 = QUAD(len=L, k1=x[3])
#     new_Q4 = QUAD(len=L, k1=x[4])
#     new_Q5 = QUAD(len=L, k1=x[5])
#     new_Q6 = QUAD(len=L, k1=x[6])
#     new_Q7 = QUAD(len=L, k1=x[7])
#     changed_ele = [new_Q1, new_Q2, new_Q3, new_Q4, new_Q5, new_Q6, new_Q7]
#     refpts = [i for i in 1:5550]
#     # refpts = [35, 5533, length(ESR_crab)]
#     twi = ADtwissring(ESR_perturb, 0.0, 0, refpts, changed_idx, changed_ele, E0=E0, m0=m_e)

#     phase41 = twi[35].mux + twi[end].mux - twi[5533].mux
#     return phase41 - 2*pi
# end

# function multi_val_op(x0, niter, step, RING)
#     target = 0.01
#     x0_vals = zeros(7, niter)
#     goal_vals = []
#     grad_vals = zeros(7, niter)
#     zero_idx = [9,13,17,23,27,31,5537]
#     g0 = get_phase14_zero(x0, ESR_perturb)
#     for i in 1:niter
#         grads = jacobian(Forward, get_phase14_zero, x0, Const(ESR_perturb))

#         x0[1] -= step * grads[1][1]
#         x0[2] -= step * grads[1][2]
#         x0[3] -= step * grads[1][3]
#         x0[4] -= step * grads[1][4]
#         x0[5] -= step * grads[1][5]
#         x0[6] -= step * grads[1][6]
#         x0[7] -= step * grads[1][7]

#         new_phase = get_phase14_zero(x0, ESR_perturb)
#         println("init: ", g0, " now: ", new_phase, "at step ", i)
#         x0_vals[:, i] = x0
#         push!(goal_vals, new_phase)
#         grad_vals[:, i] = [grads[1][1], grads[1][2], grads[1][3], grads[1][4], grads[1][5], grads[1][6], grads[1][7]]
#         if new_phase < target
#             println("tuning finished at step ", i)
#             break
#         end
#     end
#     return x0_vals, goal_vals, grad_vals
# end

# x0_vals, goal_vals, grad_vals = multi_val_op(xinit, 10, 5e-6, ESR_crab)

File: python_integration/phase_advance_tuning_demo.py
"""
This script demonstrates using pyJuTrack to tune the ESR lattice for a specific 
phase advance between two crab cavities using auto-differentiation.

Warning: This code is for demonstration purposes. Simply tuning the quadrupoles 
may result in unstable solutions.
"""

import numpy as np
import pyJuTrack as jt
import sys

E0 = 17.846262619763e9  # Beam energy in eV
jt.set_tps_dim(7) # 7 variables for the 7 quadrupoles we'll tune

# Load the ESR main lattice
# ESR_crab = jt.load_lattice("src/demo/ESR/esr_main.jls")
from juliacall import Main as jl
# RING is defined in esr_main.jl
jl.seval('include("src/demo/ESR/esr-main.jl")')
ESR_crab = jt.Lattice(jl.RING)
print(f"Loaded ESR lattice with {len(ESR_crab)} elements")
print(f"Total length: {ESR_crab.total_length():.3f} m")

ESR_crab = jt.Number2TPSAD(ESR_crab)

# Function to perturb quadrupoles to create some errors
def Q_perturb(lattice):
    """Add random perturbations to quadrupole strengths"""
    for i in range(len(lattice)):
        elem = lattice[i]
        elem_type = type(elem).__name__
        if 'KQUAD' in str(type(elem)):
            k1_val = elem.k1
            if hasattr(k1_val, 'val'):
                k1 = float(k1_val.val)
            else:
                k1 = float(k1_val)
            
            k1_perturbed = k1 * (1 + 0.001 * np.random.randn())
            
            if hasattr(k1_val, 'val'):
                new_KQ = jt.KQUAD(name=str(elem.name), length=elem.len, k1=jt.DTPSAD(k1_perturbed))
            else:
                new_KQ = jt.KQUAD(name=str(elem.name), length=float(elem.len), k1=k1_perturbed)
            lattice[i] = new_KQ
    return lattice

ESR_perturb = Q_perturb(ESR_crab)

# Find crab cavity indices
crab_idx = jt.findelem(ESR_crab, element_type=jt.element_types.CRABCAVITY)

# Indices of quadrupoles to be tuned.
# !!! Note: These are 1-based indices in Julia. Subtract 1 for 0-based Python indexing.
changed_idx = [9, 13, 17, 23, 27, 31, 5537]
print(f"\nQuadrupoles to be tuned: {changed_idx}")

# objective function
def get_phase14_zero(X):
    working_lattice = jt.Lattice(ESR_perturb.julia_object)
    
    for i, idx in enumerate(changed_idx):
        working_lattice[idx-1].k1 = X[i]
    
    refpts = list(range(1, len(working_lattice) + 1))
    twi = jt.twissring(working_lattice, dp=0.0, refpts=refpts, E0=E0, m0=jt.m_e)
    
    # Phase advance between crab cavities
    phase41 = twi[35].mux + twi[-1].mux - twi[5533].mux
    
    # Subtract 2*pi
    return phase41 - jt.DTPSAD(2.0 * np.pi)

def multi_val_optimization(x0, niter, step):
    """
    Perform gradient descent optimization to minimize phase advance error.
    
    Args:
        x0: Initial quadrupole strengths (list of 7 values)
        niter: Maximum number of iterations
        step: Step size for gradient descent
    
    Returns:
        x0_vals: History of quadrupole strengths
        goal_vals: History of objective function values
        grad_vals: History of gradients
    """
    target = 0.015  
    x0_vals = np.zeros((7, niter))
    goal_vals = []
    grad_vals = np.zeros((7, niter))
    
    grads_init, g0 = jt.Gradient(get_phase14_zero, x0, return_value=True)
    print(f"\nInitial phase advance error: {g0:.6f} rad")
    print(f"Target: {target:.6f} rad")

    for i in range(niter):
        grads, new_phase = jt.Gradient(get_phase14_zero, x0, return_value=True)
        
        for j in range(7):
            x0[j] -= step * grads[j]
        
        x0_vals[:, i] = x0
        goal_vals.append(new_phase)
        grad_vals[:, i] = grads
        
        if (i + 1) % 10 == 0 or i < 5:  # Print first 5 and every 10th iteration
            print(f"Step {i+1:3d}: Phase error = {new_phase:+.6f} rad")
        
        if abs(new_phase) < target:
            x0_vals = x0_vals[:, :i+1]
            grad_vals = grad_vals[:, :i+1]
            break
    else:
        print(f"Maximum iterations ({niter}) reached without convergence")
        print(f"Final phase error: {new_phase:.6f} rad (target: {target:.6f} rad)")
    
    return x0_vals, goal_vals, grad_vals

# Initial guess for quadrupole strengths
xinit = [-1e-6, 1e-6, -1e-6, 1e-6, -1e-6, 1e-6, -1e-6]
print(f"\nInitial quadrupole strengths: {xinit}")

# Run optimization
x0_vals, goal_vals, grad_vals = multi_val_optimization(xinit.copy(), niter=100, step=1e-4)

# Print final results
print("\nFinal quadrupole strengths:")
for i, k in enumerate(x0_vals[:, -1]):
    print(f"  k{i+1} (index {changed_idx[i]}): {k:+.6e} m^-2")
print(f"Improvement: {(goal_vals[0] - goal_vals[-1]):.6f} rad")

try:
    import matplotlib.pyplot as plt
    
    plot_steps = len(goal_vals)
    
    fig, (ax1, ax2, ax3) = plt.subplots(3, 1, figsize=(10, 10))
    
    # Plot 1: Evolution of quadrupole strengths
    for i in range(7):
        ax1.plot(range(1, plot_steps+1), x0_vals[i, :plot_steps], 
                marker='o', linestyle='--', label=f'k{i+1}', markersize=4)
    ax1.set_xlabel('Iterations')
    ax1.set_ylabel('Strength (m$^{-2}$)')
    ax1.set_title('Evolution of Quadrupole Strengths')
    ax1.legend(ncol=2)
    ax1.grid(True, alpha=0.3)
    
    # Plot 2: Evolution of phase advance error
    ax2.plot(range(1, plot_steps+1), goal_vals, 
            marker='o', linestyle='--', color='blue', linewidth=2, markersize=4)
    ax2.axhline(y=0.015, color='r', linestyle=':', linewidth=2, label='Target')
    ax2.set_xlabel('Iterations')
    ax2.set_ylabel('Phase advance error (rad)')
    ax2.set_title('Evolution of Objective Function')
    ax2.legend()
    ax2.grid(True, alpha=0.3)
    
    # Plot 3: Evolution of gradients
    for i in range(7):
        ax3.plot(range(1, plot_steps+1), grad_vals[i, :plot_steps], 
                marker='o', linestyle='--', label=f'∂Δφ/∂k{i+1}', markersize=4)
    ax3.set_xlabel('Iterations')
    ax3.set_ylabel('Gradient')
    ax3.set_title('Evolution of Gradients')
    ax3.legend(ncol=2)
    ax3.grid(True, alpha=0.3)
    
    plt.tight_layout()
    
    # Save figure
    plt.savefig('phase_advance_tuning_results.png', dpi=150, bbox_inches='tight')
    print(f"Saved plot to 'phase_advance_tuning_results.png'")
    
    # plt.show()
    
except ImportError:
    print("\nmatplotlib not available - skipping plots")
except Exception as e:
    print(f"\nError creating plots: {e}")


